spec/tapioca/compilers/dsl/smart_properties_spec.rb

```diff
@@ -0,0 +1,110 @@
+# typed: false
+# frozen_string_literal: true
+
+require "spec_helper"
+
+RSpec.describe(Tapioca::Compilers::Dsl::SmartProperties) do
+  describe("#initialize") do
+    it("gathers no constants if there are no SmartProperty classes") do
+      expect(subject.processable_constants).to(be_empty)
+    end
+
+    it("gathers only SmartProperty classes") do
+      content = <<~RUBY
+        class Post
+          include ::SmartProperties
+        end
+
+        class User
+          include ::SmartProperties
+        end
+
+        class Comment
+        end
+      RUBY
+
+      with_contents(content) do
+        expect(subject.processable_constants).to(eq(Set.new([Post, User])))
+      end
+    end
+
+    it("ignores SmartProperty classes without a name") do
+      content = <<~RUBY
+        class Post
+          include ::SmartProperties
+
+          def self.name
+            nil
+          end
+        end
+      RUBY
+
+      with_contents(content) do
+        expect(subject.processable_constants).to(be_empty)
+      end
+    end
+  end
+
+  describe("#decorate") do
+    it("generates empty RBI file if there are no smart properties") do
+      content = <<~RUBY
+        class Post
+          include SmartProperties
+        end
+      RUBY
+
+      with_contents(content) do
+        parlour = Parlour::RbiGenerator.new(sort_namespaces: true)
+        subject.decorate(parlour.root, Post)
+        expect(parlour.rbi).to(eq("# typed: strong\n\n"))
+      end
+    end
+
+    it("generates RBI file if there are smart properties") do
```

Comment:
I would really like it if we could test each combination of how a `property` could be defined individually on a separate test. That would make it much easier to identify what went wrong when something is broken.

For example, `SmartProperties` has a lot of way of defining `accepts` and `required` etc, which we had to write a lot of code to work around. Specifically, we need to carefully test the cases where `accepts` could be an array of different types, where it could have a lambda as a value, etc etc.